network/newstream/sync_provider.go

```diff
@@ -0,0 +1,359 @@
+// established only within depth ( >=depth ). This is needed for Push Sync. When set to false, the streams are
+// established on all bins as they did traditionally with Pull Sync.
+func NewSyncProvider(ns *storage.NetStore, kad *network.Kademlia, syncOnlyWithinDepth bool) *syncProvider {
+	s := &syncProvider{
+		netStore:                ns,
+		kad:                     kad,
+		syncBinsOnlyWithinDepth: syncOnlyWithinDepth,
+		name:                    syncStreamName,
+		quit:                    make(chan struct{}),
+		logger:                  log.New("base", hex.EncodeToString(kad.BaseAddr()[:8])),
+	}
+	return s
+}
+
+func (s *syncProvider) NeedData(ctx context.Context, key []byte) (loaded bool, wait func(context.Context) error) {
+	s.logger.Debug("syncProvider.NeedData", "key", hex.EncodeToString(key))
+	start := time.Now()
+
+	select {
+	case <-s.quit:
+		return false, nil
+	default:
+	}
+
+	fi, loaded, ok := s.netStore.GetOrCreateFetcher(ctx, key, "syncer")
+	if !ok {
+		s.logger.Debug("returning as if we dont need data")
+		return loaded, nil
+	}
+	s.logger.Debug("we need data", "ref", hex.EncodeToString(key))
+	return ok, func(ctx context.Context) error {
+		select {
+		case <-fi.Delivered:
+			metrics.GetOrRegisterResettingTimer(fmt.Sprintf("fetcher.%s.syncer", fi.CreatedBy), nil).UpdateSince(start)
+		case <-time.After(timeouts.SyncerClientWaitTimeout):
+			metrics.GetOrRegisterCounter("fetcher.syncer.timeout", nil).Inc(1)
+			return fmt.Errorf("chunk not delivered through syncing after %dsec. ref=%s", timeouts.SyncerClientWaitTimeout, fmt.Sprintf("%x", key))
+		}
+		return nil
+	}
+}
+
+func (s *syncProvider) Get(ctx context.Context, addr chunk.Address) ([]byte, error) {
+	log.Debug("syncProvider.Get")
+	ch, err := s.netStore.Store.Get(ctx, chunk.ModeGetSync, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// mark the chunk as Set in order to allow for garbage collection
+	// this can and at some point should be moved to a dedicated method that
+	// marks an entire sent batch of chunks as Set once the actual p2p.Send succeeds
+	err = s.netStore.Store.Set(context.Background(), chunk.ModeSetSync, addr)
+	if err != nil {
+		metrics.GetOrRegisterCounter("syncProvider.set-sync-err", nil).Inc(1)
+		return nil, err
+	}
+	return ch.Data(), nil
+}
+
+func (s *syncProvider) Put(ctx context.Context, addr chunk.Address, data []byte) (exists bool, err error) {
+	log.Trace("syncProvider.Put", "addr", addr)
+	ch := chunk.NewChunk(addr, data)
+	seen, err := s.netStore.Store.Put(ctx, chunk.ModePutSync, ch)
+	if seen {
+		log.Trace("syncProvider.Put - chunk already seen", "addr", addr)
+		if putSeenTestHook != nil {
+			// call the test function if it is set
+			putSeenTestHook(addr, s.netStore.LocalID)
+		}
+	}
+	return seen, err
+}
+
+// Function used only in tests to detect chunks that are synced
+// multiple times within the same stream. This function pointer must be
+// nil in production.
+var putSeenTestHook func(addr chunk.Address, id enode.ID)
```

Comment:
I think this is a good use-case of something we could test if we redirect logs with a simulation. Then we could check for presence of the given log message containing `addr` and `id`. For now we don't have this functionality, so fine to leave this here, but ultimately we will be able to test this kind of impl. behaviour without custom hooks.

Just an idea.